docs/security/apparmor.md

```diff
@@ -67,9 +73,111 @@ explicitly specifies the default policy:
 ```
 
+## Loading and Unloading Profiles
+
+To load a new profile into AppArmor, for use with containers:
+
+```
+$ apparmor_parser -r -W /path/to/your_profile
+```
+
+Then you can run the custom profile with `--security-opt` like so:
+
+```bash
+$ docker run --rm -it --security-opt apparmor:your_profile hello-world
+```
+
+To unload a profile from AppArmor:
+
+```bash
+# stop apparmor
+$ /etc/init.d/apparmor stop
+# unload the profile
+$ apparmor_parser -R /path/to/profile
+# start apparmor
+$ /etc/init.d/apparmor start
+```
+
+## Debugging AppArmor
+
+### Using `dmesg`
+
+Here are some helpful tips for debugging any problems you might be facing with
+regard to AppArmor.
+
+AppArmor sends quite verbose messaging to `dmesg`. Usually an AppArmor line
+will look like the following:
+
+```
+[ 5442.864673] audit: type=1400 audit(1453830992.845:37): apparmor="ALLOWED" operation="open" profile="/usr/bin/docker" name="/home/jessie/docker/man/man1/docker-attach.1" pid=10923 comm="docker" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
+```
+
+In the above example, the you can see `profile=/usr/bin/docker`. This means the
+user has the `docker-engine` (Docker Engine Daemon) profile loaded. On version
+of Ubuntu > 14.04 this is all fine and well, but Trusty users might run into
+some issues when trying to `docker exec`.
+
+Let's look at another log line:
+
+```
+[ 3256.689120] type=1400 audit(1405454041.341:73): apparmor="DENIED" operation="ptrace" profile="docker-default" pid=17651 comm="docker" requested_mask="receive" denied_mask="receive"
+```
+
+This time the profile is `docker-default`, which is run on containers by
+default unless in `privileged` mode. It is telling us, that apparmor has denied
+`ptrace` in the container. This is great.
+
+### Using `aa-status`
+
+If you need to check which profiles are loaded you can use `aa-status`. The
+output looks like:
+
+```bash
+$ sudo aa-status
+apparmor module is loaded.
+14 profiles are loaded.
+1 profiles are in enforce mode.
+   docker-default
+13 profiles are in complain mode.
+   /usr/bin/docker
+   /usr/bin/docker///bin/cat
+   /usr/bin/docker///bin/ps
+   /usr/bin/docker///sbin/apparmor_parser
+   /usr/bin/docker///sbin/auplink
+   /usr/bin/docker///sbin/blkid
+   /usr/bin/docker///sbin/iptables
+   /usr/bin/docker///sbin/mke2fs
+   /usr/bin/docker///sbin/modprobe
+   /usr/bin/docker///sbin/tune2fs
+   /usr/bin/docker///sbin/xtables-multi
+   /usr/bin/docker///sbin/zfs
+   /usr/bin/docker///usr/bin/xz
+38 processes have profiles defined.
+37 processes are in enforce mode.
+   docker-default (6044)
+   ...
+   docker-default (31899)
+1 processes are in complain mode.
+   /usr/bin/docker (29756)
+0 processes are unconfined but have a profile defined.
+```
+
+In the above output you can tell I have the `docker-default` profile running on
+various container PIDs in `enforce` mode. This means AppArmor will actively
+block and audit in `dmesg` anything outside the bounds of the `docker-default`
```

Comment:
I think `in dmesg` is usually interpreted as `in the output of dmesg`. Really it's in the kernel's message buffer, which is regular logs. I feel like any addition to this explanation gets more into the weeds than content. I'd take another opinion on how others read it.